In [5]:
import pandas as pd
df = pd.read_csv("../data/raw/NSCH 2024e_Topical_CAHMI_DRC (1).csv")
labels = pd.read_excel(
    "../data/raw/NSCH_2024e_Topical_Variable Labels_CAHMI_DRC.xlsx"
)
df.head()


FileNotFoundError: [Errno 2] No such file or directory: '../data/raw/NSCH 2024e_Topical_CAHMI_DRC (1).csv'

In [10]:
import pandas as pd

df = pd.read_csv("../data/raw/NSCH_2024_Survey_Data.csv")

labels = pd.read_excel(
    "../data/raw/NSCH_2024_Survey_Labels_ID.xlsx"
)


In [ ]:
df.shape


(51375, 902)

In [ ]:
df["HHID"].nunique()

51375

In [ ]:
len(df)

51375

In [ ]:
labels.head()

,Name,Label
0,HEIGHT,Child's Height (CM)
1,FIPSST,State FIPS Code
2,STRATUM,Sampling Stratum
3,HHID,Unique Household ID
4,FORMTYPE,Form Type


In [ ]:
labels[labels["Label"].str.contains("mental", case=False, na=False)]


,Name,Label
38,K4Q28X04,Needed Health Care Not Received - Mental Healt...
100,K2Q36A,Developmental Delay
101,K2Q36B,Developmental Delay Currently
217,ACE8,Child Experienced - Lived with Mentally Ill
237,K2Q36C,Developmental Delay Severity Description
264,A1_MENTHEALTH,Adult 1 - Mental or Emotional Health
266,A2_MENTHEALTH,Adult 2 - Mental or Emotional Health
352,K4Q22_R,Mental Health Professional Treatment
353,TREATNEED,Mental Health Professional Treatment - Problem
526,DevDelay_24,Children who currently have developmental dela...


In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51375 entries, 0 to 51374
Columns: 902 entries, HEIGHT to smForgoneHC_24
dtypes: float64(3), int64(898), str(1)
memory usage: 353.5 MB


In [ ]:
df.columns

Index(['HEIGHT', 'FIPSST', 'STRATUM', 'HHID', 'FORMTYPE', 'TOTKIDS_R',
       'TENURE', 'HHLANGUAGE', 'SC_AGE_YEARS', 'SC_SEX',
       ...
       'nomAnxietyDep12to17_24', 'nomSystCareCSHCN_24', 'nomFlrish6mto5_24',
       'nomFlrish6to17_24', 'nomFlrish6to17CSHCN_24', 'nomACE2more_24',
       'smSmoking_24', 'smPhysAct12to17_24', 'smAdeqIns_24', 'smForgoneHC_24'],
      dtype='str', length=902)

In [ ]:
labels.info()

<class 'pandas.DataFrame'>
RangeIndex: 902 entries, 0 to 901
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Name    902 non-null    str  
 1   Label   902 non-null    str  
dtypes: str(2)
memory usage: 14.2 KB


In [ ]:
labels[labels["Label"].str.contains("insurance", case=False, na=False)]

,Name,Label
162,KEEPINSADULT,How to Obtain/Keep Health Insurance as Child B...
163,CURRCOV,Health Insurance Coverage - Currently Covered ...
164,K12Q03,Health Insurance - Current/Former Employer or ...
165,K12Q04,Health Insurance - Insurance Company
166,K12Q12,Health Insurance - Government Assistance Plan
167,TRICARE,Health Insurance - TRICARE
168,K11Q03R,Health Insurance - Indian Health Service
169,HCCOVOTH,Health Insurance - Other
173,AVOIDCHG,Past 12 Months - Avoided Changing Jobs to Main...
277,K3Q20,Health Insurance - Benefits Cover Services


In [ ]:
labels[labels["Label"].str.contains("poverty", case=False, na=False)]

,Name,Label
450,FPL_I1,"Family Poverty Ratio, First Implicate"
451,FPL_I2,"Family Poverty Ratio, Second Implicate"
452,FPL_I3,"Family Poverty Ratio, Third Implicate"
453,FPL_I4,"Family Poverty Ratio, Fourth Implicate"
454,FPL_I5,"Family Poverty Ratio, Fifth Implicate"
455,FPL_I6,"Family Poverty Ratio, Sixth Implicate"
468,povlev4_24,Income level based on family poverty level sta...
469,povSCHIP_24,Income level based on family poverty level sta...
780,WrkngPoorR_24,Indicator 6.5a: Children living in 'working po...


In [ ]:
labels[labels["Label"].str.contains("anxiety", case=False, na=False)]

,Name,Label
84,K2Q33A,Anxiety
85,K2Q33B,Anxiety Currently
232,K2Q33C,Anxiety Severity Description
523,anxiety_24,"Children who currently have anxiety problems, ..."
547,AnxietSev_24,Parent-rated severity of child's current anxie...
892,nomAnxietyDep12to17_24,"NOM: Percent of adolescents, ages 12 through 1..."


In [ ]:
labels.to_csv("../docs/NSCH_2024_Survey_Labels_ID.csv", index=False)

In [ ]:
import pandas as pd

df = pd.read_csv("../data/raw/NSCH_2024_Survey_Data.csv")

print("Dataset Shape:", df.shape)
df.head()

Dataset Shape: (51375, 902)


,HEIGHT,FIPSST,STRATUM,HHID,FORMTYPE,TOTKIDS_R,TENURE,HHLANGUAGE,SC_AGE_YEARS,SC_SEX,...,nomAnxietyDep12to17_24,nomSystCareCSHCN_24,nomFlrish6mto5_24,nomFlrish6to17_24,nomFlrish6to17CSHCN_24,nomACE2more_24,smSmoking_24,smPhysAct12to17_24,smAdeqIns_24,smForgoneHC_24
0,9999.0,48,1,24000016,3,3,2,1,12,2,...,2,2,90,99,99,2,2,2,1,2
1,9990.0,48,1,24000017,1,1,2,1,5,2,...,95,95,1,90,90,2,2,90,1,2
2,9990.0,47,1,24000044,1,2,1,1,5,1,...,95,95,1,90,90,2,2,90,1,2
3,9990.0,13,2,24000054,1,2,2,1,2,2,...,95,95,1,90,90,2,2,90,1,2
4,9990.0,39,1,24000057,1,2,1,1,2,2,...,95,95,1,90,90,2,2,90,1,2


# NSCH 2024 Data Cleaning & Exploration

This notebook prepares the National Survey of Children's Health (NSCH) 2024 dataset for analysis. The objective is to assess data quality, identify missing values and duplicates, verify data types, and produce an analysis-ready dataset for studying unmet pediatric mental health needs.


In [ ]:
# Dataset dimensions
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")
df.info()
df.describe()
df.describe(include="all")
missing = df.isnull().sum()

missing = missing[missing > 0].sort_values(ascending=False)

missing.head(25)
missing_percent = (
    df.isnull()
      .mean()
      .mul(100)
      .sort_values(ascending=False)
)

missing_percent.head(25)
duplicates = df.duplicated().sum()

print(f"Duplicate Rows: {duplicates}")

Rows: 51,375
Columns: 902
<class 'pandas.DataFrame'>
RangeIndex: 51375 entries, 0 to 51374
Columns: 902 entries, HEIGHT to smForgoneHC_24
dtypes: float64(3), int64(898), str(1)
memory usage: 353.5 MB
Duplicate Rows: 0


In [ ]:
df.dtypes.value_counts()

int64      898
float64      3
str          1
Name: count, dtype: int64

# Survey Code Exploration

In [ ]:
# Check height values
df["HEIGHT"].value_counts(dropna=False).head(15)
# Check age values
df["SC_AGE_YEARS"].value_counts(dropna=False)
# Check sex values
df["SC_SEX"].value_counts(dropna=False)

SC_SEX
1    26374
2    25001
Name: count, dtype: int64

In [ ]:
df.info()
duplicates = df.duplicated().sum()
print(duplicates)
df["HEIGHT"].value_counts().head(20)

<class 'pandas.DataFrame'>
RangeIndex: 51375 entries, 0 to 51374
Columns: 902 entries, HEIGHT to smForgoneHC_24
dtypes: float64(3), int64(898), str(1)
memory usage: 353.5 MB
0


HEIGHT
9990.00000    13841
162.56000      2336
157.48000      2010
152.39999      2002
121.92000      1949
167.64000      1936
160.02000      1841
165.10001      1806
170.17999      1633
172.72000      1506
154.94000      1424
175.25999      1310
9999.00000     1299
149.86000      1258
147.32001      1245
177.80000      1210
137.16000      1141
142.24001      1059
127.00000      1041
144.78000       935
Name: count, dtype: int64

# Height inspection


In [ ]:
import pandas as pd

# Load the labels file
labels = pd.read_excel("../data/raw/NSCH_2024_Survey_Labels_ID.xlsx")

# Search every column for "HEIGHT"
labels[labels.astype(str)
       .apply(lambda x: x.str.contains("HEIGHT", case=False, na=False))
       .any(axis=1)]

,Name,Label
0,HEIGHT,Child's Height (CM)


In [ ]:
df["HEIGHT"].describe()
df["HEIGHT"].sort_values().tail(20)

34044    9999.0
47393    9999.0
47525    9999.0
34168    9999.0
12912    9999.0
47398    9999.0
34336    9999.0
34319    9999.0
12890    9999.0
25836    9999.0
47520    9999.0
47434    9999.0
8593     9999.0
8629     9999.0
47498    9999.0
8657     9999.0
25903    9999.0
1909     9999.0
1944     9999.0
0        9999.0
Name: HEIGHT, dtype: float64

In [ ]:
sorted(df["HEIGHT"].dropna().unique())[-20:]


[np.float64(147.32001),
 np.float64(149.86),
 np.float64(152.39999),
 np.float64(154.94),
 np.float64(157.48),
 np.float64(160.02),
 np.float64(162.56),
 np.float64(165.10001),
 np.float64(167.64),
 np.float64(170.17999),
 np.float64(172.72),
 np.float64(175.25999),
 np.float64(177.8),
 np.float64(180.34),
 np.float64(182.88),
 np.float64(185.42),
 np.float64(187.96001),
 np.float64(193.75999),
 np.float64(9990.0),
 np.float64(9999.0)]

In [ ]:
df["HEIGHT"].dropna().unique()

array([9999.     , 9990.     ,  137.16   ,  175.25999,  139.7    ,
        162.56   ,  152.39999,  182.88   ,  160.02   ,  170.17999,
        144.78   ,  167.64   ,  103.93   ,  180.34   ,  187.96001,
        177.8    ,  147.32001,  149.86   ,  165.10001,  114.3    ,
        142.24001,  132.08   ,  121.92   ,  111.76   ,  172.72   ,
        127.     ,  154.94   ,  109.22   ,  134.62   ,  157.48   ,
        129.53999,  124.46   ,  193.75999,  119.38   ,  185.42   ,
        116.84   ])

In [ ]:
sorted(df["HEIGHT"].dropna().unique())

[np.float64(103.93),
 np.float64(109.22),
 np.float64(111.76),
 np.float64(114.3),
 np.float64(116.84),
 np.float64(119.38),
 np.float64(121.92),
 np.float64(124.46),
 np.float64(127.0),
 np.float64(129.53999),
 np.float64(132.08),
 np.float64(134.62),
 np.float64(137.16),
 np.float64(139.7),
 np.float64(142.24001),
 np.float64(144.78),
 np.float64(147.32001),
 np.float64(149.86),
 np.float64(152.39999),
 np.float64(154.94),
 np.float64(157.48),
 np.float64(160.02),
 np.float64(162.56),
 np.float64(165.10001),
 np.float64(167.64),
 np.float64(170.17999),
 np.float64(172.72),
 np.float64(175.25999),
 np.float64(177.8),
 np.float64(180.34),
 np.float64(182.88),
 np.float64(185.42),
 np.float64(187.96001),
 np.float64(193.75999),
 np.float64(9990.0),
 np.float64(9999.0)]

# Define Analysis Set

In [ ]:
analysis_vars = [
    # Outcomes
    "MentHCare_24",
    "MentHDiff_24",

    # Demographics
    "SC_AGE_YEARS",
    "SC_SEX",
    "SC_RACE_R",
    "HISPANIC",

    # Socioeconomic
    "povlev4_24",
    "FPL_IF",
    "TENURE",
    "ACE1",
    "FOODSIT",
    "MISSMORTGAGE",
    "EVERHOMELESS",

    # Access to Care
    "PrevMed_24",
    "K4Q20R",
    "K4Q04_R",
    "HOWMUCH",

    # Mental Health
    "ADHDMed_24",
    "ADHDBehTreat_24",

    # School
    "GRADES",
    "MAKEFRIEND",

    # Neighborhood
    "K10Q30",
    "K10Q31",
    "K10Q40_R",
    "K10Q41_R",

    # Family
    "K8Q11",
    "K8Q21",
    "K8Q30"
]

In [ ]:
analysis_df = df[analysis_vars].copy()

print(analysis_df.shape)
analysis_df.head()

KeyError: "['HISPANIC'] not in index"

In [ ]:
missing_vars = [col for col in analysis_vars if col not in df.columns]

print("Variables not found:")
for col in missing_vars:
    print(col)

Variables not found:
HISPANIC


# Race / Ethnicity

In [ ]:
[col for col in df.columns if "RACE" in col.upper()]

['SC_RACE_R',
 'SC_RACE_R_IF',
 'SC_RACER',
 'race4_24',
 'raceAsia_24',
 'race7_24']

In [ ]:
[col for col in df.columns if "HISP" in col.upper()]

['SC_HISPANIC_R', 'SC_HISPANIC_R_IF', 'hisplang_24']

In [ ]:
[col for col in df.columns if "DEPR" in col.upper()]

['depress_24', 'DepresSev_24']

# Mental Health

In [ ]:
[col for col in df.columns if "MENT" in col.upper()]

['APPOINTMENT',
 'A1_MENTHEALTH',
 'A2_MENTHEALTH',
 'MentHCare_24',
 'MentHDiff_24',
 'appointment_24',
 'mentor_24',
 'EmploymentSt_24',
 'npmMentHTx12to17_24',
 'npmMentor12to17_24']

In [ ]:
[col for col in df.columns if "ANXI" in col.upper()]

['anxiety_24', 'AnxietSev_24', 'nomAnxietyDep12to17_24']

In [ ]:
[col for col in df.columns if "DEPR" in col.upper()]

['depress_24', 'DepresSev_24']

# Poverty

In [ ]:
[col for col in df.columns if "POV" in col.upper()]

['povlev4_24', 'povSCHIP_24']

In [ ]:
[col for col in df.columns if "FPL" in col.upper()]

['FPL_IF', 'FPL_I1', 'FPL_I2', 'FPL_I3', 'FPL_I4', 'FPL_I5', 'FPL_I6']

# Insurance 

In [ ]:
[col for col in df.columns if "INS" in col.upper()]

['KEEPINSADULT',
 'GAINSKILLS',
 'CURRINS',
 'INSGAP',
 'INSTYPE',
 'CurrIns_24',
 'InsGap_24',
 'instype_24',
 'instype3_24',
 'InsAdeq_24',
 'insurance_24',
 'DentInstruct_24',
 'GainSkill_24',
 'InsNeeds_24',
 'SmkInside_24',
 'HousingInstab_24',
 'WCRIdomains_24',
 'npmHousingInstab_24',
 'smAdeqIns_24']

# ACEs

In [ ]:
[col for col in df.columns if "ACE" in col.upper()]

['ACE3',
 'ACE4',
 'ACE5',
 'ACE6',
 'ACE7',
 'ACE8',
 'ACE9',
 'ACE10',
 'ACE11',
 'DRAWAFACE',
 'ACE1',
 'SC_RACE_R',
 'SC_RACE_R_IF',
 'SC_RACER',
 'race4_24',
 'raceAsia_24',
 'race7_24',
 'DrawFace_24',
 'ACEct_24',
 'ACE2more_24',
 'ACEincome_24',
 'ACEincome2_24',
 'ACEdivorce_24',
 'ACEdeath_24',
 'ACEjail_24',
 'ACEdomviol_24',
 'ACEneighviol_24',
 'ACEmhealth_24',
 'ACEdrug_24',
 'ACEdiscrim_24',
 'ACEHealthDiscrim_24',
 'ACE6ctHH_24',
 'ACE2more6HH_24',
 'ACEctComm_24',
 'ACE1moreComm_24',
 'nomACE2more_24']

## Variable Discovery

The NSCH dataset contains 902 variables, so only variables relevant to the research question will be selected for analysis.

This section searches the dataset for variables related to:

- Mental health care and access
- Anxiety and depression
- Poverty and household resources
- Insurance coverage
- Race and ethnicity
- Adverse Childhood Experiences (ACEs)
- School, family, and neighborhood factors

Exact dataset column names are verified before creating the analysis dataset.

In [ ]:
def search_variables(keyword):
    """
    Return dataset column names containing a keyword.

    Parameters
    ----------
    keyword : str
        Word or partial word to search for in column names.
    """
    matches = [
        column
        for column in df.columns
        if keyword.upper() in column.upper()
    ]

    print(f"Found {len(matches)} variables containing '{keyword}':\n")

    for column in matches:
        print(column)

    return matches

### Mental Health

In [ ]:
mental_health_vars = search_variables("MENT")

Found 10 variables containing 'MENT':

APPOINTMENT
A1_MENTHEALTH
A2_MENTHEALTH
MentHCare_24
MentHDiff_24
appointment_24
mentor_24
EmploymentSt_24
npmMentHTx12to17_24
npmMentor12to17_24


### Anxiety

In [ ]:
anxiety_vars = search_variables("ANXI")

Found 3 variables containing 'ANXI':

anxiety_24
AnxietSev_24
nomAnxietyDep12to17_24


### Depression

In [ ]:
depression_vars = search_variables("DEPR")

Found 2 variables containing 'DEPR':

depress_24
DepresSev_24


### ADHD 

In [ ]:
[col for col in df.columns if "ADHD" in col.upper()]

['ADHD_24',
 'ADHDSev_24',
 'ADHDind_24',
 'ADHDSevInd_24',
 'ADHDMed_24',
 'ADHDBehTreat_24']

In [ ]:
adhd_vars = search_variables("ADHD")

Found 6 variables containing 'ADHD':

ADHD_24
ADHDSev_24
ADHDind_24
ADHDSevInd_24
ADHDMed_24
ADHDBehTreat_24


### Poverty

In [ ]:
poverty_vars = search_variables("POV")
fpl_vars = search_variables("FPL")

Found 2 variables containing 'POV':

povlev4_24
povSCHIP_24
Found 7 variables containing 'FPL':

FPL_IF
FPL_I1
FPL_I2
FPL_I3
FPL_I4
FPL_I5
FPL_I6


### Insurance

In [ ]:
insurance_vars = search_variables("INS")

Found 19 variables containing 'INS':

KEEPINSADULT
GAINSKILLS
CURRINS
INSGAP
INSTYPE
CurrIns_24
InsGap_24
instype_24
instype3_24
InsAdeq_24
insurance_24
DentInstruct_24
GainSkill_24
InsNeeds_24
SmkInside_24
HousingInstab_24
WCRIdomains_24
npmHousingInstab_24
smAdeqIns_24


### Race and Ethnicity

In [ ]:
race_vars = search_variables("RACE")
hispanic_vars = search_variables("HISP")

Found 6 variables containing 'RACE':

SC_RACE_R
SC_RACE_R_IF
SC_RACER
race4_24
raceAsia_24
race7_24
Found 3 variables containing 'HISP':

SC_HISPANIC_R
SC_HISPANIC_R_IF
hisplang_24


In [ ]:
hispanic_vars = search_variables("HISP")

### Adverse Childhood Experiences

Adverse Childhood Experiences, or ACEs, are potentially stressful or traumatic experiences that may affect a child's health and well-being.

The NSCH dataset contains individual ACE items and derived summary measures, including counts and indicators for two or more ACEs.

In [ ]:
ace_vars = search_variables("ACE")

Found 36 variables containing 'ACE':

ACE3
ACE4
ACE5
ACE6
ACE7
ACE8
ACE9
ACE10
ACE11
DRAWAFACE
ACE1
SC_RACE_R
SC_RACE_R_IF
SC_RACER
race4_24
raceAsia_24
race7_24
DrawFace_24
ACEct_24
ACE2more_24
ACEincome_24
ACEincome2_24
ACEdivorce_24
ACEdeath_24
ACEjail_24
ACEdomviol_24
ACEneighviol_24
ACEmhealth_24
ACEdrug_24
ACEdiscrim_24
ACEHealthDiscrim_24
ACE6ctHH_24
ACE2more6HH_24
ACEctComm_24
ACE1moreComm_24
nomACE2more_24


In [ ]:
ace_vars = [
    column
    for column in df.columns
    if column.upper().startswith("ACE")
    or column.upper().startswith("NOMACE")
]

print(f"Found {len(ace_vars)} ACE-related variables:\n")

for column in ace_vars:
    print(column)

Found 28 ACE-related variables:

ACE3
ACE4
ACE5
ACE6
ACE7
ACE8
ACE9
ACE10
ACE11
ACE1
ACEct_24
ACE2more_24
ACEincome_24
ACEincome2_24
ACEdivorce_24
ACEdeath_24
ACEjail_24
ACEdomviol_24
ACEneighviol_24
ACEmhealth_24
ACEdrug_24
ACEdiscrim_24
ACEHealthDiscrim_24
ACE6ctHH_24
ACE2more6HH_24
ACEctComm_24
ACE1moreComm_24
nomACE2more_24


### School

In [ ]:
school_vars = search_variables("SCHOOL")
attendance_vars = search_variables("ATTEND")
bully_vars = search_variables("BULL")

Found 3 variables containing 'SCHOOL':

STARTSCHOOL
SchoolReadiness_24
nomSchoolReadiness2_24
Found 0 variables containing 'ATTEND':

Found 8 variables containing 'BULL':

BULLIED_R
BULLY
bully_24
bullied_24
npmBully12to17_24
npmBully12to17CSHCN_24
npmBullied12to17_24
npmBullied12to17CSHCN_24


### Family

In [ ]:
family_vars = search_variables("FAMILY")
parent_vars = search_variables("PARENT")

Found 1 variables containing 'FAMILY':

FAMILY_R
Found 3 variables containing 'PARENT':

ParentPhH_24
ParentMH_24
ParentHSt_24


### Neighborhood & Community

In [ ]:
neighborhood_vars = search_variables("NEIGH")
community_vars = search_variables("COMM")
safety_vars = search_variables("SAFE")

Found 1 variables containing 'NEIGH':

ACEneighviol_24
Found 5 variables containing 'COMM':

DrCommR_24
OtherComm_24
SatOthComm_24
ACEctComm_24
ACE1moreComm_24
Found 2 variables containing 'SAFE':

NbhdSafe_24
SchlSafe_24


### Sleep

In [ ]:
sleep_vars = search_variables("SLEEP")

Found 3 variables containing 'SLEEP':

HOURSLEEP05
HOURSLEEP
HrsSleep_24


### Physical Activity

In [ ]:
exercise_vars = search_variables("EXER")
activity_vars = search_variables("ACTIVE")
sports_vars = search_variables("SPORT")

Found 2 variables containing 'EXER':

ENGAGE_EXERCISE
OverExercise_24
Found 3 variables containing 'ACTIVE':

A1_ACTIVE
A2_ACTIVE
ActiveWork_24
Found 3 variables containing 'SPORT':

TRANSPORTCC
transport_24
sports_24


### Gneral Health

In [ ]:
health_vars = search_variables("HEALTH")

Found 8 variables containing 'HEALTH':

HEALTHKNOW
A1_PHYSHEALTH
A1_MENTHEALTH
A2_PHYSHEALTH
A2_MENTHEALTH
mhealth_24
ACEmhealth_24
ACEHealthDiscrim_24


### Special Health Care Needs

In [ ]:
special_health_vars = search_variables("SHCN")
special_needs_vars = search_variables("SPECIAL")

Found 19 variables containing 'SHCN':

TOTCSHCN
TOTNONSHCN
SC_CSHCN
CSHCN_24
CSHCNtype_24
CSHCNComplex_24
CSHCNExpanded_24
CSHCNComplexExp_24
npmMedHomeCSHCN_24
npmPerDrNsCSHCN_24
npmUsualSckCSHCN_24
npmFamCentCSHCN_24
npmNoRefPrb2CSHCN_24
npmCareCoorRCSHCN_24
npmTransitionCSHCN_24
npmBully12to17CSHCN_24
npmBullied12to17CSHCN_24
nomSystCareCSHCN_24
nomFlrish6to17CSHCN_24
Found 0 variables containing 'SPECIAL':



### Household

In [ ]:
household_vars = search_variables("HOUSE")
housing_vars = search_variables("HOUS")

Found 1 variables containing 'HOUSE':

HOUSE_GEN
Found 4 variables containing 'HOUS':

HOUSE_GEN
HousingInstab_24
housing_24
npmHousingInstab_24


# Create the Analysis-Ready Dataset

The full NSCH dataset contains 902 variables and children ages 0–17. This project focuses on school-aged children ages 6–17 and variables relevant to unmet pediatric mental health needs.

The cleaned dataset will retain:

- Survey design and identification variables
- Child demographics
- Mental health conditions and severity
- Mental health care access
- Insurance and healthcare access
- Poverty and material hardship
- Adverse Childhood Experiences
- Parent and family characteristics
- School and peer experiences
- Neighborhood conditions
- Sleep, physical activity, and screen time
- Special healthcare needs

Official NSCH derived measures are used whenever possible. Special survey codes representing legitimate skips, children outside a question’s age range, suppressed responses, or missing responses will be converted to missing values.

In [16]:
selected_vars = [

    # ==========================================
    # Survey Design Variables
    # ==========================================

    "HHID",
    "STRATUM",
    "FWC",  # Final survey sampling weight, not the child's physical weight

    # ==========================================
    # Geography & Built Environment
    # ==========================================

    # Geographic Identifiers
    "FIPSST",
    "MPC_YN",
    "hrsareg",

    # Neighborhood Environment
    "NbhdSupp_24",
    "NbhdSafe_24",
    "NbhdAmenities_24",

    # Community Environment
    "OtherComm_24",
    "DrCommR_24",

    # Built Environment
    "SideWlks_24",
    "park_24",
    "RecCentr_24",
    "library_24",

    # ==========================================
    # Demographics
    # ==========================================

    "SC_AGE_YEARS",
    "SC_SEX",
    "SC_RACE_R",
    "SC_HISPANIC_R",

    # ==========================================
    # Mental Health & Neurodevelopmental Conditions
    # ==========================================

    # Anxiety
    "anxiety_24",
    "AnxietSev_24",

    # Depression
    "depress_24",
    "DepresSev_24",

    # ADHD
    "ADHD_24",
    "ADHDSev_24",
    "ADHDind_24",
    "ADHDSevInd_24",

    # Autism Spectrum Disorder
    "autism_24",
    "AutismSev_24",
    "AutismInd_24",

    # Behavioral / Conduct Problems
    "behavior_24",
    "BehavSev_24",

    # Tourette Syndrome
    "tourette_24",
    "TouretSev_24",

    # Developmental Conditions
    "DevDelay_24",
    "speech_24",
    "learning_24",
    "IntDisab_24",

    # ==========================================
    # Mental Health Treatment & Healthcare Access
    # ==========================================

    # ADHD Treatment
    "ADHDMed_24",
    "ADHDBehTreat_24",

    # Autism Treatment
    "ASDMed_24",
    "ASDBehTreat_24",

    # General Mental Health Treatment
    "MedEmotion_24",
    "therapy_24",
    "rxmeds_24",

    # Mental Health Care Access
    "MentHCare_24",
    "MentHDiff_24",

    # ==========================================
    # Mental Health Indicators & Composite Measures
    # ==========================================

    "mhealth_24",
    "mhneeds_24",
    "nomAnxietyDep12to17_24",
    "MEDB10ScrQ5_24",

    # ==========================================
    # Socioeconomic Status
    # ==========================================

    "povlev4_24",
    "povSCHIP_24",
    "FPL_IF",

    # ==========================================
    # Insurance & Healthcare Access
    # ==========================================

    "insurance_24",
    "CurrIns_24",
    "InsGap_24",
    "InsAdeq_24",

    # ==========================================
    # Adverse Childhood Experiences (ACEs)
    # ==========================================

    "ACEct_24",
    "ACE2more_24",

    # ==========================================
    # Family & Household Environment
    # ==========================================

    "FAMILY_R",
    "ParentPhH_24",
    "ParentMH_24",
    "ParentHSt_24",

    # ==========================================
    # School Readiness & Development
    # ==========================================

    "SchoolReadiness_24",

    "EL_domain_24",
    "SE_domain_24",
    "SR_domain_24",
    "M_domain_24",
    "H_domain_24",

    "flrish6to17_24",
    "flrsh6to17ct",

    # ==========================================
    # School Experience & Academic Performance
    # ==========================================

    "SchlEngage_24",
    "cares_24",
    "GRADES",
    "ReptGrade_24",
    "SchlMiss_24",

    # ==========================================
    # School Social Environment
    # ==========================================

    "bullied_24",
    "bully_24",
    "SchlSafe_24",

    # ==========================================
    # Extracurricular & Community Participation
    # ==========================================

    "AftSchAct_24",
    "sports_24",
    "clubs_24",
    "lessons_24",
    "volunteer_24",
    "EventPart_24",

    # ==========================================
    # Housing Stability
    # ==========================================

    "HousingInstab_24",

    # ==========================================
    # Physical Health & Lifestyle
    # ==========================================

    "WEIGHT",  # Child's current physical weight in kilograms
    "HrsSleep_24",
    "ActiveWork_24",

    # ==========================================
    # Special Health Care Needs
    # ==========================================

    "CSHCN_24",
    "CSHCNComplex_24",

]

## Dataset Validation

Before constructing the curated dataset, all selected variables were validated against the original NSCH dataset to confirm that each variable name was correct and available for analysis.

Duplicate variables and missing variable names were checked and resolved before dataset creation.

In [17]:
# Validate selected variable names against the original dataset

missing_vars = [
    var
    for var in selected_vars
    if var not in df.columns
]
print(f"Variables selected: {len(selected_vars)}")
print(f"Variables found: {len(selected_vars) - len(missing_vars)}")
print(f"Variables missing: {len(missing_vars)}")

if missing_vars:
    print("\nMissing variables:")
    for var in missing_vars:
        print(f"- {var}")
else:
    print("\nAll selected variables were found in the dataset.")

Variables selected: 92
Variables found: 92
Variables missing: 0

All selected variables were found in the dataset.


In [18]:
# Check whether any variables were selected more than once

duplicate_vars = sorted({
    var
    for var in selected_vars
    if selected_vars.count(var) > 1
})

print(f"Duplicate variables: {len(duplicate_vars)}")

if duplicate_vars:
    for var in duplicate_vars:
        print(f"- {var}")
else:
    print("No duplicate variables were found.")

Duplicate variables: 0
No duplicate variables were found.


In [19]:
# Create curated dataset using validated variables

curated_df = df[selected_vars].copy()

print("Curated dataset created.")
print(f"Rows: {curated_df.shape[0]:,}")
print(f"Columns: {curated_df.shape[1]}")

Curated dataset created.
Rows: 51,375
Columns: 92


In [20]:
# Create curated dataset from validated variables

curated_df = df[selected_vars].copy()

print("Curated dataset created.")
print(f"Rows: {curated_df.shape[0]:,}")
print(f"Columns: {curated_df.shape[1]}")

Curated dataset created.
Rows: 51,375
Columns: 92


In [21]:
display(curated_df.head())

,HHID,STRATUM,FWC,FIPSST,MPC_YN,hrsareg,NbhdSupp_24,NbhdSafe_24,NbhdAmenities_24,OtherComm_24,...,clubs_24,lessons_24,volunteer_24,EventPart_24,HousingInstab_24,WEIGHT,HrsSleep_24,ActiveWork_24,CSHCN_24,CSHCNComplex_24
0,24000016,1,14108.624370,48,2,6,1,1,99,2,...,2,2,2,1,2,9999.0,1,1,1,2
1,24000017,1,7995.775609,48,2,6,1,1,4,95,...,90,90,90,90,2,9990.0,1,90,2,3
2,24000044,1,1593.130507,47,2,4,1,1,1,2,...,90,90,90,90,2,9990.0,1,90,2,3
3,24000054,2,4395.677115,13,2,4,1,1,1,2,...,90,90,90,90,2,9990.0,1,90,2,3
4,24000057,1,1384.962653,39,2,5,1,1,1,2,...,90,90,90,90,2,9990.0,1,90,2,3


## Variable and Value Labeling

Original NSCH variable names will be preserved in the curated dataset for traceability.

A separate cleaned analysis dataset will use readable column names and labeled response values for use in Power BI, visualization, and interpretation.

In [22]:
# Check the available ages before filtering

curated_df["SC_AGE_YEARS"].value_counts(dropna=False).sort_index()

SC_AGE_YEARS
0     1537
1     1949
2     2870
3     2462
4     2489
5     2534
6     2203
7     2408
8     2563
9     2626
10    2860
11    2914
12    2923
13    3165
14    3459
15    3691
16    4288
17    4434
Name: count, dtype: int64

In [23]:
# Filter to the study population: children ages 6–17

analysis_df = curated_df[
    curated_df["SC_AGE_YEARS"].between(6, 17)
].copy()

print("Age filter complete.")
print(f"Rows before filter: {len(curated_df):,}")
print(f"Rows after filter: {len(analysis_df):,}")
print(f"Rows removed: {len(curated_df) - len(analysis_df):,}")
print(f"Age range: {analysis_df['SC_AGE_YEARS'].min()}–{analysis_df['SC_AGE_YEARS'].max()}")

Age filter complete.
Rows before filter: 51,375
Rows after filter: 37,534
Rows removed: 13,841
Age range: 6–17


## Survey Code Audit

Before recoding response values, the unique values and frequencies for each selected variable were reviewed.

This step identifies valid response categories, age-ineligible responses, skipped questions, refusals, unknown responses, and other survey-specific codes. No original values were replaced until their meanings were confirmed.

In [29]:
# Audit each variable for missing values, unique values,
# and common NSCH special response codes

special_codes = [90, 95, 96, 97, 98, 99]

audit_rows = []

for column in analysis_df.columns:
    value_counts = analysis_df[column].value_counts(
        dropna=False
    )

    audit_rows.append({
        "variable": column,
        "data_type": str(analysis_df[column].dtype),
        "rows": len(analysis_df),
        "missing_nan": analysis_df[column].isna().sum(),
        "unique_values": analysis_df[column].nunique(dropna=True),
        "code_90": (analysis_df[column] == 90).sum(),
        "code_95": (analysis_df[column] == 95).sum(),
        "code_96": (analysis_df[column] == 96).sum(),
        "code_97": (analysis_df[column] == 97).sum(),
        "code_98": (analysis_df[column] == 98).sum(),
        "code_99": (analysis_df[column] == 99).sum(),
    })

code_audit_df = pd.DataFrame(audit_rows)

display(code_audit_df)

,variable,data_type,rows,missing_nan,unique_values,code_90,code_95,code_96,code_97,code_98,code_99
0,HHID,int64,37534,0,37534,0,0,0,0,0,0
1,STRATUM,int64,37534,0,2,0,0,0,0,0,0
2,FWC,float64,37534,0,33037,0,0,0,0,0,0
3,FIPSST,int64,37534,0,51,0,0,0,0,0,0
4,MPC_YN,int64,37534,0,3,0,0,5215,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...
87,WEIGHT,float64,37534,0,179,0,0,0,0,0,0
88,HrsSleep_24,int64,37534,0,3,0,0,0,0,0,495
89,ActiveWork_24,int64,37534,0,4,15574,0,0,0,0,2352
90,CSHCN_24,int64,37534,0,2,0,0,0,0,0,0


In [27]:
labels_df = pd.read_csv("../docs/NSCH_2024_Survey_Labels_ID.csv")

print(labels_df.shape)
display(labels_df.head())

(902, 2)


,Name,Label
0,HEIGHT,Child's Height (CM)
1,FIPSST,State FIPS Code
2,STRATUM,Sampling Stratum
3,HHID,Unique Household ID
4,FORMTYPE,Form Type


In [31]:
# Rebuild and label the special-code audit in one cell

special_code_columns = [
    "code_90",
    "code_95",
    "code_96",
    "code_97",
    "code_98",
    "code_99",
]

special_code_audit = code_audit_df.loc[
    code_audit_df[special_code_columns].sum(axis=1) > 0
].copy()

labels_lookup = labels_df.rename(
    columns={
        "Name": "variable",
        "Label": "variable_label",
    }
)

special_code_audit_labeled = special_code_audit.merge(
    labels_lookup,
    on="variable",
    how="left",
)

display(
    special_code_audit_labeled[
        [
            "variable",
            "variable_label",
            "code_90",
            "code_95",
            "code_96",
            "code_97",
            "code_98",
            "code_99",
        ]
    ]
)

,variable,variable_label,code_90,code_95,code_96,code_97,code_98,code_99
0,MPC_YN,Metropolitan Principal City Status,0,0,5215,0,0,0
1,NbhdSupp_24,Indicator 7.1: Children who live in supportive...,0,0,0,0,0,1280
2,NbhdSafe_24,Indicator 7.2: Children live in safe neighborh...,0,0,0,0,0,1047
3,NbhdAmenities_24,Indicator 7.4: Count of amenities in children'...,0,0,0,0,0,1272
4,OtherComm_24,Health care provider communicate with child’s ...,0,5400,0,0,0,449
...,...,...,...,...,...,...,...,...
67,volunteer_24,Indicator 5.7: Children participated in commun...,0,0,0,0,0,727
68,EventPart_24,Indicator 5.6: Parents participated in child’s...,0,0,0,0,0,522
69,HousingInstab_24,Indicator 6.29: Children who experienced housi...,0,0,0,0,0,715
70,HrsSleep_24,Indicator 6.25: Child slept recommended age-ap...,0,0,0,0,0,495


In [32]:
# Inspect workbook sheets to see whether value labels are included

xlsx_path = "../data/raw/NSCH_2024_Survey_Labels_ID.xlsx"

excel_file = pd.ExcelFile(xlsx_path)

print(excel_file.sheet_names)

['Sheet1']


### Special Code Documentation Gap

The available NSCH variable-label files include variable names and descriptions but do not define the meanings of special response codes such as 90, 95, 96, and 99.

These codes were therefore audited and preserved until the official NSCH codebook or value-label documentation could be consulted.

## Readable Variable Names

The original NSCH variable names are preserved in `analysis_df` for traceability.

A separate cleaned dataset is created with readable snake_case column names for analysis, visualization, and Power BI.

In [33]:
cleaned_df = analysis_df.copy()

In [34]:
rename_map = {
    # Survey design
    "HHID": "household_id",
    "STRATUM": "sampling_stratum",
    "FWC": "final_sampling_weight",

    # Geography
    "FIPSST": "state_fips_code",
    "MPC_YN": "metropolitan_principal_city_status",
    "hrsareg": "hrsa_region",

    # Neighborhood
    "NbhdSupp_24": "neighborhood_support",
    "NbhdSafe_24": "neighborhood_safety",
    "NbhdAmenities_24": "neighborhood_amenities_count",

    # Community
    "OtherComm_24": "provider_communication_with_other_services",
    "DrCommR_24": "provider_communication_rating",

    # Built environment
    "SideWlks_24": "sidewalks_available",
    "park_24": "park_available",
    "RecCentr_24": "recreation_center_available",
    "library_24": "library_available",

    # Demographics
    "SC_AGE_YEARS": "child_age_years",
    "SC_SEX": "child_sex",
    "SC_RACE_R": "child_race",
    "SC_HISPANIC_R": "child_hispanic_ethnicity",
}

In [35]:
cleaned_df = cleaned_df.rename(columns=rename_map)

print(f"Columns renamed: {len(rename_map)}")
display(cleaned_df.head())

Columns renamed: 19


,household_id,sampling_stratum,final_sampling_weight,state_fips_code,metropolitan_principal_city_status,hrsa_region,neighborhood_support,neighborhood_safety,neighborhood_amenities_count,provider_communication_with_other_services,...,clubs_24,lessons_24,volunteer_24,EventPart_24,HousingInstab_24,WEIGHT,HrsSleep_24,ActiveWork_24,CSHCN_24,CSHCNComplex_24
0,24000016,1,14108.624370,48,2,6,1,1,99,2,...,2,2,2,1,2,9999.000000,1,1,1,2
7,24000081,1,1763.619834,47,2,4,2,2,1,2,...,2,2,2,1,2,28.120001,2,90,2,3
8,24000088,1,109.931073,31,2,7,1,1,3,2,...,1,1,1,1,2,95.250000,2,2,1,1
9,24000089,2,948.037623,27,2,5,1,1,2,2,...,1,1,2,2,2,29.480000,1,90,2,3
17,24000173,1,532.319837,9,2,1,1,1,3,1,...,1,1,99,2,2,49.900002,2,99,2,3


In [36]:
print("Original columns preserved:")
print("SC_AGE_YEARS" in analysis_df.columns)

print("\nReadable columns created:")
print("child_age_years" in cleaned_df.columns)

Original columns preserved:
True

Readable columns created:
True


In [37]:
stata_path = "../data/raw/nsch_2024e_topical.dta"

stata_df = pd.read_stata(
    stata_path,
    convert_categoricals=True
)

print(stata_df.shape)
display(stata_df.head())

(51375, 457)


,height,fipsst,stratum,hhid,formtype,totkids_r,tenure,hhlanguage,sc_age_years,sc_sex,...,hhcount_if,higrade,higrade_tvis,fpl_i1,fpl_i2,fpl_i3,fpl_i4,fpl_i5,fpl_i6,fwc
0,NaN,48,1,24000016,T3,3,2,1.0,12,2,...,0,3,4,400,400,400,400,400,400,14108.624367
1,NaN,48,1,24000017,T1,1,2,1.0,5,2,...,0,3,4,400,400,400,400,400,400,7995.775609
2,NaN,47,1,24000044,T1,2,1,1.0,5,1,...,0,3,4,400,400,400,400,239,195,1593.130507
3,NaN,13,2A,24000054,T1,2,2,1.0,2,2,...,0,3,4,400,400,400,400,400,400,4395.677115
4,NaN,39,1,24000057,T1,2,1,1.0,2,2,...,0,3,4,400,400,400,400,400,400,1384.962653


In [38]:
display(stata_df["HrsSleep_24"].value_counts(dropna=False))
display(stata_df["MPC_YN"].value_counts(dropna=False))
display(stata_df["OtherComm_24"].value_counts(dropna=False))
display(stata_df["ActiveWork_24"].value_counts(dropna=False))

KeyError: 'HrsSleep_24'

In [39]:
display(stata_df["hrssleep_24"].value_counts(dropna=False))
display(stata_df["mpc_yn"].value_counts(dropna=False))
display(stata_df["othercomm_24"].value_counts(dropna=False))
display(stata_df["activework_24"].value_counts(dropna=False))

KeyError: 'hrssleep_24'

In [40]:
matches = [
    col
    for col in stata_df.columns
    if any(
        term in col.lower()
        for term in [
            "sleep",
            "mpc",
            "othercomm",
            "activework",
        ]
    )
]

matches

['hoursleep05', 'hoursleep', 'mpc_yn']

In [41]:
display(stata_df["hoursleep05"].value_counts(dropna=False))
display(stata_df["hoursleep"].value_counts(dropna=False))
display(stata_df["mpc_yn"].value_counts(dropna=False))

hoursleep05
NaN    37666
7.0     4384
5.0     3458
6.0     2932
4.0     1673
3.0      997
2.0      208
1.0       57
Name: count, dtype: int64

hoursleep
NaN    14336
4.0    13073
5.0    10341
6.0     5959
3.0     4963
2.0     1370
7.0      951
1.0      382
Name: count, dtype: int64

mpc_yn
2.0    31121
1.0    13116
NaN     7138
Name: count, dtype: int64

In [42]:
[col for col in stata_df.columns if "other" in col.lower()]

['blood_other', 'visionscreenother']

In [43]:
# Search Stata columns and labels for the derived sleep indicator

reader = pd.io.stata.StataReader(stata_path)

variable_labels = reader.variable_labels()

sleep_matches = {
    name: label
    for name, label in variable_labels.items()
    if "recommended" in label.lower()
    or "sleep" in label.lower()
}

sleep_matches

{'hoursleep05': 'HOURSLEEP05', 'hoursleep': 'HOURSLEEP'}

In [44]:
# Inspect all Stata value-label dictionaries

value_labels = reader.value_labels()

value_labels.keys()

dict_keys([])

In [45]:
{
    name: mapping
    for name, mapping in value_labels.items()
    if any(
        term in name.lower()
        for term in ["sleep", "mpc"]
    )
}

{}

In [46]:
variable_value_labels = reader.variable_to_label()

{
    variable: label_set
    for variable, label_set in variable_value_labels.items()
    if variable in ["hoursleep05", "hoursleep", "mpc_yn"]
}

AttributeError: 'StataReader' object has no attribute 'variable_to_label'

In [47]:
# Read the Stata setup file as plain text

do_file_path = "../data/raw/nsch_2024_topical.do"

with open(do_file_path, "r", encoding="utf-8", errors="ignore") as file:
    do_text = file.read()

print(f"Characters read: {len(do_text):,}")
print(do_text[:2000])

Characters read: 221,075
/* 06062025 */
*************************************************************
*************************************************************
local file = "<<INSERT FILE DIRECTORY>>"
*************************************************************
*************************************************************

use "`file'\nsch_2024e_topical", clear

label var fpl_if  "Imputation Flag for FPL"
label var a1_grade_if  "Imputation Flag for A1_GRADE"
label var sc_hispanic_r_if  "Imputation Flag for SC_HISPANIC_R"
label var sc_race_r_if  "Imputation Flag for SC_RACE_R"
label var sc_sex_if  "Imputation Flag for SC_SEX"
label var hhcount_if  "Imputation Flag for HHCOUNT"
label var insgap  "Health Insurance Coverage - Past 12 Months"
label var instype  "Insurance Type (Revised)"
label var currins  "Health Insurance Coverage - Currently Covered"
label var birthwt  "Birth Weight Status"
label var mpc_yn  "Metropolitan Principal City Status"
label var metro_yn  "Metropolitan Stat

In [48]:
# Show lines in the Stata setup file that mention special codes

special_code_lines = [
    line.strip()
    for line in do_text.splitlines()
    if any(
        code in line
        for code in ["90", "95", "96", "97", "98", "99"]
    )
]

print(f"Matching lines: {len(special_code_lines)}")

for line in special_code_lines[:200]:
    print(line)

Matching lines: 12
label var k9q96  "Other Adult Child Can Rely On For Advice"
label define bmiclass_lab  3  "85th percentile to less than the 95th percentile", add
label define bmiclass_lab  4  "Equal to or greater than the 95th percentile", add
label define howmuch_lab  3  "$250-$499", add
label define howmuch_lab  4  "$500-$999", add
label define k9q96_lab  1  "Yes"
label define k9q96_lab  2  "No", add
label define k9q96_lab  .m "No valid response", add
label define k9q96_lab  .n "Not in universe", add
label define k9q96_lab  .l "Logical skip", add
label define k9q96_lab  .d "Suppressed for confidentiality", add
cap label values k9q96 k9q96_lab


In [49]:
# Reload Stata data while preserving its distinct missing-value codes

stata_missing_df = pd.read_stata(
    stata_path,
    convert_categoricals=False,
    convert_missing=True
)

print(stata_missing_df.shape)

(51375, 457)


In [50]:
stata_missing_df["mpc_yn"].value_counts(dropna=False)

mpc_yn
2     31121
1     13116
.d     7138
Name: count, dtype: int64

In [51]:
stata_missing_meanings = {
    ".m": "No valid response",
    ".n": "Not in universe",
    ".l": "Logical skip",
    ".d": "Suppressed for confidentiality",
}

In [52]:
for value, count in stata_missing_df["mpc_yn"].value_counts(
    dropna=False
).items():
    print(
        f"value={value!s}, "
        f"type={type(value).__name__}, "
        f"count={count:,}"
    )

value=2, type=int, count=31,121
value=1, type=int, count=13,116
value=.d, type=StataMissingValue, count=7,138


In [53]:
mpc_status_map = {
    1: "Yes",
    2: "No",
    96: "Suppressed for confidentiality",
}

In [54]:
# Prepare Stata columns for matching with the CSV

stata_compare_df = stata_missing_df.copy()

# Convert Stata column names to uppercase so they resemble CSV names
stata_compare_df.columns = [
    col.upper()
    for col in stata_compare_df.columns
]

# Confirm the household ID is available
print("HHID" in analysis_df.columns)
print("HHID" in stata_compare_df.columns)

True
True


In [56]:
# Convert Stata special missing values to readable strings before grouping

mpc_compare["MPC_YN_stata_label"] = (
    mpc_compare["MPC_YN_stata"]
    .map(str)
)

mpc_mapping = (
    mpc_compare
    .groupby(
        ["MPC_YN_csv", "MPC_YN_stata_label"],
        dropna=False,
    )
    .size()
    .reset_index(name="rows")
)

display(mpc_mapping)

,MPC_YN_csv,MPC_YN_stata_label,rows
0,1,1,9377
1,2,2,22942
2,96,.d,5215


In [58]:
csv_special_codes = [90, 95, 96, 97, 98, 99]

mapping_rows = []

common_columns = [
    col
    for col in analysis_df.columns
    if col in stata_compare_df.columns
    and col != "HHID"
]

for column in common_columns:
    comparison = analysis_df[
        ["HHID", column]
    ].merge(
        stata_compare_df[
            ["HHID", column]
        ],
        on="HHID",
        how="inner",
        suffixes=("_csv", "_stata"),
    )

    comparison = comparison[
        comparison[f"{column}_csv"].isin(csv_special_codes)
    ]

    for raw_code, stata_value in comparison[
        [f"{column}_csv", f"{column}_stata"]
    ].itertuples(index=False, name=None):

        mapping_rows.append({
            "variable": column,
            "csv_code": raw_code,
            "stata_value": str(stata_value),
        })

special_code_mapping = pd.DataFrame(mapping_rows)

if special_code_mapping.empty:
    print("No overlapping special-code mappings were found.")
else:
    special_code_mapping = (
        special_code_mapping
        .groupby(
            ["variable", "csv_code", "stata_value"],
            dropna=False,
        )
        .size()
        .reset_index(name="rows")
        .sort_values(["variable", "csv_code", "stata_value"])
    )

    display(special_code_mapping)

,variable,csv_code,stata_value,rows
0,FAMILY_R,99,.m,904
1,GRADES,99,.m,571
2,MPC_YN,96,.d,5215
